// @blockfrost/openapi/lib/functions/schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateSchema = exports.getSchema = exports.generateSchemas = exports.getSchemaForEndpoint = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const yaml_1 = __importDefault(require("yaml"));
const ajv_1 = __importDefault(require("ajv"));
const nutlink_address_tickers_1 = __importDefault(require("../custom-schemas/nutlink-address-tickers"));
const nutlink_ticker_1 = __importDefault(require("../custom-schemas/nutlink-ticker"));
const scripts_json_1 = __importDefault(require("../custom-schemas/scripts-json"));
const txs_metadata_1 = __importDefault(require("../custom-schemas/txs-metadata"));
const ajv = new ajv_1.default({ strict: false });
const file = fs_1.default.readFileSync(path_1.default.resolve(__dirname, '../../openapi.yaml'), 'utf8');
const spec = yaml_1.default.parse(file);
const getSchemaForEndpoint = (endpointName) => {
    if (!spec.paths[endpointName]) {
        throw Error(`Missing Blockfrost OpenAPI schema for endpoint "${endpointName}".`);
    }
    const responses = { response: {} };
    // Hacky way to support POST endpoints.
    // Primarily pick GET with a fallback to POST
    // TODO: return body of POST endpoints
    // https://www.fastify.io/docs/latest/Reference/Validation-and-Serialization/#validation
    const method = 'post' in spec.paths[endpointName] ? 'post' : 'get';
    const endpoint = spec.paths[endpointName][method];
    for (const response of Object.keys(endpoint.responses)) {
        // success 200
        if (response === '200') {
            const contentType = 'application/octet-stream' in endpoint.responses['200'].content
                ? 'application/octet-stream'
                : 'application/json';
            const referenceOrValue = endpoint.responses['200'].content[contentType].schema;
            // is reference -> resolve references
            if (referenceOrValue['$ref']) {
                const schemaName = referenceOrValue['$ref'].replace('#/components/schemas/', '');
                const schemaReferenceOrValue = spec.components.schemas[schemaName];
                // is nested reference
                if (schemaReferenceOrValue.items &&
                    schemaReferenceOrValue.items['$ref']) {
                    const nestedSchemaName = schemaReferenceOrValue.items['$ref'].replace('#/components/schemas/', '');
                    if (schemaReferenceOrValue.type) {
                        responses.response[200] = {
                            ...schemaReferenceOrValue,
                            items: spec.components.schemas[nestedSchemaName],
                        };
                    }
                    else {
                        responses.response[200] = spec.components.schemas[nestedSchemaName];
                    }
                }
                else {
                    // is not nested reference
                    responses.response[200] = spec.components.schemas[schemaName];
                }
            }
            else {
                // is not reference
                responses.response[200] = referenceOrValue;
            }
            // anyOf case
            if (referenceOrValue['anyOf']) {
                const anyOfResult = { anyOf: [] };
                for (const anyOfItem of referenceOrValue['anyOf']) {
                    const schemaName = anyOfItem['$ref'].replace('#/components/schemas/', '');
                    const item = spec.components.schemas[schemaName];
                    anyOfResult['anyOf'].push(item);
                }
                responses.response[200] = anyOfResult;
            }
            const parameters = endpoint.parameters;
            if (parameters) {
                const queryParams = parameters.filter((i) => i.in === 'query');
                let queryProps = {};
                if (queryParams && queryParams.length > 0) {
                    queryParams.forEach((param) => {
                        delete param.schema.format;
                        queryProps[param.name] = param.schema;
                    });
                    responses['querystring'] = {
                        type: 'object',
                        properties: queryProps,
                    };
                }
                const pathParams = parameters.filter((i) => i.in === 'path');
                if (pathParams && pathParams.length > 0) {
                    let pathProps = {};
                    pathParams.forEach((param) => {
                        delete param.schema.format;
                        pathProps[param.name] = param.schema;
                    });
                    responses['params'] = {
                        type: 'object',
                        properties: pathProps,
                    };
                }
                // const query = parameters.filter((i: any) => i.in === 'param');
                // let queryParams: any = {};
            }
            // custom schemas
            if (endpointName === '/txs/{hash}/metadata') {
                responses.response[200] = txs_metadata_1.default;
            }
            if (endpointName === '/nutlink/{address}/tickers/{ticker}') {
                responses.response[200] = nutlink_address_tickers_1.default;
            }
            if (endpointName === '/nutlink/tickers/{ticker}') {
                responses.response[200] = nutlink_ticker_1.default;
            }
            if (endpointName === '/scripts/{script_hash}/json') {
                // TODO: no longer necessary
                responses.response[200] = scripts_json_1.default;
            }
        }
        // errors and others
        else {
            responses.response[response] =
                spec.components.responses[response].content['application/json'].schema;
        }
    }
    // debug
    // if (endpointName === '/blocks/{hash_or_number}') {
    //   console.log(endpointName, JSON.stringify(responses));
    // }
    return responses;
};
exports.getSchemaForEndpoint = getSchemaForEndpoint;
const generateSchemas = () => {
    // Returns fast-json-stringify compatible schema object indexed by endpoint name
    const endpoints = Object.keys(spec.paths);
    const schemas = {};
    for (const endpoint of endpoints) {
        try {
            schemas[endpoint] = (0, exports.getSchemaForEndpoint)(endpoint);
        }
        catch (error) {
            console.error(`Error while processing endpoint ${endpoint}`);
            throw error;
        }
    }
    return schemas;
};
exports.generateSchemas = generateSchemas;
const getSchema = (schemaName) => {
    if (!spec.components.schemas[schemaName]) {
        throw Error(`Missing Blockfrost OpenAPI schema with name "${schemaName}".`);
    }
    return spec.components.schemas[schemaName];
};
exports.getSchema = getSchema;
const validateSchema = (schemaName, input) => {
    const schema = (0, exports.getSchema)(schemaName);
    const validate = ajv.compile(schema);
    const isValid = validate(input);
    return { isValid, errors: validate.errors };
};
exports.validateSchema = validateSchema;
